src/llvm-late-gc-lowering.cpp

```diff
@@ -0,0 +1,1351 @@
+                                if (DT.dominates(Lifetime, SI))
+                                    AnyDominates = true;
+                            }
+                            if (AnyDominates)
+                                NoteUse(S, BBS, SI->getValueOperand());
+                        }, V);
+
+                        continue;
+                    }
+                }
+                int SafepointNumber = NoteSafepoint(S, BBS, CI);
+                BBS.HasSafepoint = true;
+                BBS.TopmostSafepoint = SafepointNumber;
+                BBS.Safepoints.push_back(SafepointNumber);
+            } else if (LoadInst *LI = dyn_cast<LoadInst>(&I)) {
+                // Allocas get sunk into the gc frame, so don't generate defs
+                MaybeNoteDef(S, BBS, LI, BBS.Safepoints);
+                NoteOperandUses(S, BBS, I, BBS.UpExposedUsesUnrooted);
+            } else if (SelectInst *SI = dyn_cast<SelectInst>(&I)) {
+                // We need to insert an extra select for the GC root
+                if (!isSpecialPtr(SI->getType()))
+                    continue;
+                if (getValueAddrSpace(SI) != AddressSpace::Tracked) {
+                    if (S.AllPtrNumbering.find(SI) != S.AllPtrNumbering.end())
+                        continue;
+                    LiftSelect(S, SI);
+                } else {
+                    MaybeNoteDef(S, BBS, SI, BBS.Safepoints);
+                    NoteOperandUses(S, BBS, I, BBS.UpExposedUsesUnrooted);
+                }
+            } else if (PHINode *Phi = dyn_cast<PHINode>(&I)) {
+                // We can have phis of jlcall frames. We consider each of the
+                // values stored to that jlcall frame a phi use
+                if (!isSpecialPtr(Phi->getType())) {
+                    MaybeNotePhiJLCallFrameUses(S, BBS, Phi);
+                    continue;
+                }
+                // We need to insert an extra phi for the GC root
+                if (getValueAddrSpace(Phi) != AddressSpace::Tracked) {
+                    if (S.AllPtrNumbering.find(Phi) != S.AllPtrNumbering.end())
+                        continue;
+                    LiftPhi(S, Phi);
+                } else {
+                    MaybeNoteDef(S, BBS, Phi, BBS.Safepoints);
+                    for (unsigned i = 0; i < Phi->getNumIncomingValues(); ++i) {
+                        BBState &IncombingBBS = S.BBStates[Phi->getIncomingBlock(i)];
+                        NoteUse(S, IncombingBBS, Phi->getIncomingValue(i), IncombingBBS.PhiOuts);
+                    }
+                }
+            } else if (isa<StoreInst>(&I) || isa<ReturnInst>(&I)) {
+                NoteOperandUses(S, BBS, I, BBS.UpExposedUsesUnrooted);
+            } else if (auto *ASCI = dyn_cast<AddrSpaceCastInst>(&I)) {
+                if (getValueAddrSpace(ASCI) == AddressSpace::Tracked)
+                    MaybeNoteDef(S, BBS, ASCI, BBS.Safepoints);
```

Comment:
It's not a safepoint, but I do treat it as a potential def here. In the future, I want to be a be more aggressive and not number these, but for now this is safer.